### DISCUSSION TITLE
Morph Target in  Model Viewer

### DISCUSSION THREAD

**Discussion by Dovatech** [google/model-viewer#2614] [Ideas]
Hello , is it possiable to use MorphTarget Method from Three.js  in Model Viewer ? https://threejs.org/examples/webgl_morphtargets

- model-viewer: vX.X.X

### Browser Affected

- [x] Chrome, version: xx.x.xxxx.xx
- [x] Edge
- [x] Firefox
- [ ] IE
- [ ] Safari

### OS

- [ ] Android
- [ ] iOS
- [ ] Linux
- [ ] MacOS
- [x] Windows

### AR

- [x] WebXR
- [ ] SceneViewer
- [ ] QuickLook

**Comment by ghost** [comment]
MorphTarget animations are supported, I don't understand exactly what
functionality you are looking for, do you want to procedurally apply two
morph targets?

**Comment by Dovatech** [comment]
Yes. I would like to change the 3D object using Shape Keys, as in the example ( https://threejs.org/examples/webgl_morphtargets ).

**Comment by ghost** [comment]
It looks like you need access to mesh.morphTargetInfluences (
https://threejs.org/docs/#api/en/objects/Mesh.morphTargetInfluences). This
is not exposed in ModelViewer, however, I'm in the process of exposing
several other features in the scene, we can take a look at this one too.

**Comment by Dovatech** [comment]
Yes. This is exactly what I need. It would be great if this could be applied in Model Viewer. 🙏

**Comment by elalish** [comment]
Can you give us some details of your use case? You can select a given morph target animation currently. You can also hack down into our three.js layer and have full control of animation mixers, but I'm not sure we want to expose that level of complexity in our high-level API.

**Comment by Dovatech** [comment]
I have a wardrobe model with different doors and drawers. The user must be able to open and close them. Animation is not very suitable for me here, because the user can open several cells at the same time, and several animations cannot work at the same time.

**Comment by elalish** [comment]
Okay, thanks. The best I can offer at the moment is to animate an open one closing before animating the next one to open. We'll consider what kind of API might be a simple approach for the more general problem.

**Comment by adam-code-labx** [comment]
Hi @elalish
We have recently starting using model-viewer for our avatar configurator project and this has really helped us so far with our project.
Especially with the built in AR implementation 👍 

We are facing the same issue with not be able to use shape keys with our avatars.
It would be great if we were able to expose morphTargetInfluences for the shape keys.

Do you guys have any ETA for this feature?
Alternatively if you can point me in the right direction on how to expose the API to this method I would be more than happy once we have a working solution our side to submit a PR.
Thanks.

**Reply by elalish** [reply]
PRs always welcome. What do you mean by a shape key and how would you use it? Morph target animations are supported; is it that you need a full animation mixer?

**Reply by adam-code-labx** [reply]
Thanks @elalish 
We are not looking at this stage to directly animate our character. This will certainly be required at a later stage in our project.
Currently we are able to change the colour of skin using `body.pbrMetallicRoughness.setBaseColorFactor`

Our next step is to change different parts of our characters using morph sliders.
For example we would like to achieve some of the results in this example https://blog.farazshaikh.com/stories/character-creation-in-three-js/ however we do not need the dat.GUI lib as we will be controlling the sliders from our own UI

I believe a good place to start would be extending `packages/model-viewer/src/features/scene-graph/three-dom-element.ts`
Then updating the `ThreeDOMElementMap` in `packages/model-viewer/src/features/scene-graph/api.ts`

Could you please clarify on the process to extend the scene-graph api and the build process to activate the feature in our custom branch? Thanks.

**Reply by elalish** [reply]
So, a Shape Key equals a Morph Target, which is one of two types of animation in glTF. You can have a bunch of different named animations and choose one at a time right now. I believe the only thing we're missing is the ability to arbitrarily combine these different animations (which the three.js animation mixer supports). Is that what you want? This sounds more like an extension to the animation API than the scene-graph API to me. 

If you could sketch out a potential API shape for us to review, that would be a great place to start.

**Reply by adam-code-labx** [reply]
I will put together a basic avatar three.js codesandbox tomorrow so you can see exactly what we are trying to achieve.
Then we can work out and perhaps collaborate on what we will need to expose for the scene-graph api or as you said perhaps extend the animation API.
Thanks.

**Reply by adam-code-labx** [reply]
Not sure why but having issues setting this up with codesandbox very strange invalid json for my GLB so decided to create quick loom video as this is working under my local env. https://www.loom.com/share/7aa21ba1310843d59571ea93a95988f0

This is an extremely basic example using 2 morph sliders.
I believe we will only need access to `morphTargetDictionary` & `morphTargetInfluences`
From what I can see here, this should not require extending the animation API just exposing Object3D & Mesh should be enough?

Please advice.
Thanks.

**Reply by elalish** [reply]
Exposing `Object3D` and `Mesh` will expose a large fraction of the three.js API, which is not our intention here. Keep in mind, you can always hack your way down to our `$scene` private member if you want three's full API, we just don't promise anything about breaking changes down there. 

Our API works at the level of glTF's structure, so I think animation is the right way to think of this (each slider is a `currentTime` of its referenced animation name). Still, this could easily snowball into a large API surface, so we should tread carefully.

**Reply by adam-code-labx** [reply]
Thanks @elalish 
I have managed to access the three js scene via the getOwnPropertySymbols function.

For example I can now retrieve all the morphTargets
Perhaps this is a hack, however I can imagine via the API we can expose the threeScene only if required for advanced use.
something like `modelViewer.threeScene`

This is now working from our side using the following allows us to get our morphTargets and from there we are able to update the morphTargetInfluences values.

```
const model = modelViewer.model;
const threeScene = model[Object.getOwnPropertySymbols(model)[7]];
threeScene.traverse((child) => {
    if (child.isMesh) {
        child.material.metalness = 0;
        child.material.vertexColors = false;

        if (child.morphTargetDictionary) {
            for (const key in child.morphTargetDictionary) {
                const index = child.morphTargetDictionary[key];
                if (Array.isArray(morphTargets[key])) {
                    morphTargets[key].push({index, child});
                } else {
                    morphTargets[key] = [];
                    morphTargets[key].push({index, child});
                }
            }
        }
    }
});

console.info(morphTargets);
```
Thanks for all your help let me know if you would like me to create a small PR to expose threeScene for advanced use?

**Reply by elalish** [reply]
Thanks, but no thanks. I don't intend to mark that scene as part of our public API, as I have no way to ensure there aren't breaking changes there. Advanced users are always welcome to hack into it, but keeping it behind a symbol at least reminds them that they'll need to do more testing on each version update.

**Reply by adam-code-labx** [reply]
Ok I understand thanks anyway.
